src/modules/roc_netio/target_libuv/roc_netio/udp_sender_port.cpp

```diff
@@ -143,6 +151,13 @@ void UdpSenderPort::write(const packet::PacketPtr& pp) {
         roc_panic("udp sender: unexpected packet w/o data");
     }
 
+    if (rate_limiter_.allow()) {
+        const double nb_ratio =
+            nb_packet_counter_ != 0 ? (double)nb_packet_counter_ / packet_counter_ : 0.;
+        roc_log(LogDebug, "udp sender: total=%u non-blocking=%u nb_ratio=%.5f",
+                packet_counter_, nb_packet_counter_, nb_ratio);
+    }
```

Comment:
We should move it under the lock, otherwise we'll get a data race on packet_counter_.

Regarding threads in roc_netio.

UdpSenderPort::write() is called from pipeline thread, which performs sound processing and sound I/O. It case of sender, it gets sound from audio file or device, processes samples, formats packets, and passes them to UdpSenderPort.

UdpSenderPort::write_cb_() and other callbacks are called from libuv thread, which performs network I/O.

uv_async_send(), called from write() is a thread-safe way to wake up libuv from another thread. It schedules write_cb_() invocation on libuv thread. Under the hood, uv_async_send() uses eventfd or similar thing.